### ISSUE TITLE
Compatibility issues in trace v1.24.0

### ISSUE THREAD

**Issue by TomasPhilippart** [open-telemetry/opentelemetry-go#5227]
### Description

I encountered compatibility issues while using the OpenTelemetry SDK with my Go application. The error messages indicate type mismatches and missing method implementations related to the tracing package. Detailed error message (occurs during go build):

```sh
# go.opentelemetry.io/otel/sdk/trace
/Users/user/go/pkg/mod/go.opentelemetry.io/otel/sdk@v1.24.0/trace/span.go:165:20: cannot use (*recordingSpan)(nil) (value of type *recordingSpan) as ReadWriteSpan value in variable declaration: *recordingSpan does not implement ReadWriteSpan (missing method AddLink)
/Users/user/go/pkg/mod/go.opentelemetry.io/otel/sdk@v1.24.0/trace/span.go:804:20: cannot use nonRecordingSpan{} (value of type nonRecordingSpan) as "go.opentelemetry.io/otel/trace".Span value in variable declaration: nonRecordingSpan does not implement "go.opentelemetry.io/otel/trace".Span (missing method AddLink)
/Users/user/go/pkg/mod/go.opentelemetry.io/otel/sdk@v1.24.0/trace/tracer.go:50:21: impossible type assertion: p.(*recordingSpan)
	*recordingSpan does not implement "go.opentelemetry.io/otel/trace".Span (missing method AddLink)
/Users/user/go/pkg/mod/go.opentelemetry.io/otel/sdk@v1.24.0/trace/tracer.go:120:10: cannot use tr.newNonRecordingSpan(sc) (value of type nonRecordingSpan) as "go.opentelemetry.io/otel/trace".Span value in return statement: nonRecordingSpan does not implement "go.opentelemetry.io/otel/trace".Span (missing method AddLink)
/Users/user/go/pkg/mod/go.opentelemetry.io/otel/sdk@v1.24.0/trace/tracer.go:122:9: cannot use tr.newRecordingSpan(psc, sc, name, samplingResult, config) (value of type *recordingSpan) as "go.opentelemetry.io/otel/trace".Span value in return statement: *recordingSpan does not implement "go.opentelemetry.io/otel/trace".Span (missing method AddLink)

Compilation finished with exit code 1
```

### Environment

- OS: macOS 14.4.1
- Architecture: arm64
- Go Version: 1.28.1 darwin/arm64
- opentelemetry-go version: v1.24.0

### Expected behavior

No errors.

**Comment by MrAlias** [comment]
This looks like a partial upgrade. You are using the v1.25.0 API somewhere.

Upgrade everything to v1.25.0 and this should resolve your issue.

Please post your `go.mod` file if that doesn't resolve your issue.

**Comment by TomasPhilippart** [comment]
Solved by downgrading to v1.24.0:

```sh
$ go get go.opentelemetry.io/otel/trace@v1.24.0

go: downgraded go.opentelemetry.io/otel v1.25.0 => v1.24.0
go: downgraded go.opentelemetry.io/otel/metric v1.25.0 => v1.24.0
go: downgraded go.opentelemetry.io/otel/trace v1.25.0 => v1.24.0
```
